File: 351002/Bubich/t310/app/repository/base_repository.py
from abc import ABC, abstractmethod
from typing import List, Optional


class BaseRepository(ABC):
    @abstractmethod
    def save(self, entity) -> object:
        pass

    @abstractmethod
    def find_by_id(self, id: int) -> Optional[object]:
        pass

    @abstractmethod
    def find_all(self) -> List[object]:
        pass

    @abstractmethod
    def update(self, entity) -> object:
        pass

    @abstractmethod
    def delete_by_id(self, id: int) -> bool:
        pass

File: 351004/Leshok/Lab6/repositories/BaseRepository.js
// repositories/BaseRepository.js
class BaseRepository {
  constructor(store, idGenerator) {
    this.store = store;
    this.idGenerator = idGenerator;
  }

  findAll() {
    return Array.from(this.store.values());
  }

  findById(id) {
    return this.store.get(Number(id));
  }

  create(entity) {
    const id = this.idGenerator();
    const newEntity = { ...entity, id };
    this.store.set(id, newEntity);
    return newEntity;
  }

  update(id, data) {
    const existing = this.findById(id);
    if (!existing) return null;
    const updated = { ...existing, ...data, id: existing.id };
    this.store.set(id, updated);
    return updated;
  }

  delete(id) {
    return this.store.delete(Number(id));
  }
}

module.exports = BaseRepository;